velox/common/base/ScopedLock.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <mutex>
+#include <vector>
+
+namespace facebook {
+namespace velox {
+
+/// This template class is a lock wrapper which is similar to std::lock_guard
+/// and additionally, it allows user to add deferred callbacks to invoke right
+/// after the lock release. The following is the code snippets to show how to
+/// use it:
+///    std::mutex mu;
+///    {
+///       ScopedLock sl(&mu);
+///       sl.add([&]() {
+///          std::lock_guard l(mu);
+///          printf("acquired lock again"); });
+///    }
+template <typename LockType>
+class ScopedLock {
+ public:
+  ScopedLock(LockType* lock) : lock_(lock) {
+    lock_->lock();
```

Comment:
what happens if lock() throws here? looks like we're still going to call unlock() and run the callbacks. Is this the expected behavior? Would be nice to clarify in the comments above.

Reply:
We don't expect lock/unlock throws here. I will comment that behavior clearly. Thanks!

Reply:
But it could (I don't think lock is a "nothrow" function)